core/peerset/src/lib.rs

```diff
@@ -103,130 +199,127 @@ pub struct PeersetConfig {
-
-		} else {
-			// All slots are filled with reserved peers.
-			if inner.discovered.iter().all(|p| *p != peer_id) {
-				inner.discovered.push(peer_id);
+		match self.data.out_slots.add_peer(peer_id.clone(), SlotType::Reserved) {
+			Ok(_) => {
+				self.data.discovered.remove_peer(&peer_id);
+				self.message_queue.push_back(Message::Connect(peer_id));
+			},
+			Err(SlotError::AlreadyConnected(_)) => {
+				return;
+			}
+			Err(SlotError::MaxConnections(_)) => {
+				self.data.discovered.add_peer(peer_id, SlotType::Reserved);
+			}
+			Err(SlotError::DemandReroute { disconnect, ..}) => {
+				self.message_queue.push_back(Message::Drop(disconnect));
+				self.data.discovered.add_peer(peer_id, SlotType::Reserved);
 			}
 		}
 	}
 
-	/// Remove a previously-added reserved peer.
-	///
-	/// Has no effect if the node was not a reserved peer.
-	pub fn remove_reserved_peer(&self, peer_id: &PeerId) {
-		let mut inner = self.inner.lock();
-		inner.reserved.remove(peer_id);
+	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
+		self.data.in_slots.mark_not_reserved(&peer_id);
+		self.data.out_slots.mark_not_reserved(&peer_id);
+		self.data.discovered.mark_not_reserved(&peer_id);
+		if self.data.reserved_only {
+			if self.data.in_slots.contains(&peer_id) || self.data.out_slots.contains(&peer_id) {
+				self.message_queue.push_back(Message::Drop(peer_id))
+			}
+		}
 	}
 
-	/// Sets whether or not the peerset only has connections .
-	pub fn set_reserved_only(&self, reserved_only: bool) {
-		let mut inner = self.inner.lock();
-		let inner = &mut *inner;	// Fixes a borrowing issue.
-		inner.reserved_only = reserved_only;
-
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		// Disconnect non-reserved nodes.
-		if reserved_only {
-			for slot in inner.out_slots.iter_mut().chain(inner.in_slots.iter_mut()) {
-				if let Some(peer) = slot.as_ref() {
-					if inner.reserved.contains(peer) {
-						continue;
-					}
-
-					let _ = self.tx.unbounded_send(Message::Drop(peer.clone()));
-				}
-
-				*slot = None;
+		self.data.reserved_only = reserved_only;
+		if self.data.reserved_only {
+			for peer in self.data.in_slots.common_peers().chain(self.data.out_slots.common_peers()) {
+				// peer will be removed from `in_slots` or `out_slots` in `on_dropped` method
+				self.message_queue.push_back(Message::Drop(peer.clone()))
 			}
 		}
 	}
 
-	/// Reports an adjustement to the reputation of the given peer.
-	pub fn report_peer(&self, _peer_id: &PeerId, _score_diff: i32) {
-		// This is not implemented in this dummy implementation.
-	}
-}
-
-fn alloc_slots(inner: &mut Inner, tx: &mpsc::UnboundedSender<Message>) {
-	if inner.reserved_only {
-		return;
-	}
+	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
+		let score = self.data.scores.entry(peer_id.clone()).or_default();
+		*score = score.saturating_add(score_diff);
 
-	for slot in inner.out_slots.iter_mut() {
-		if slot.is_some() {
-			continue;
+		if *score < 0 {
```

Comment:
that might be overly aggressive if we start at `0`.

Reply:
This is just an arbitrary number. Once we define the pricing, we can adjust this value